Add GSoC 2021 intro post. (#21)
ruma/ruma.github.io: content/news/2021-07-15-gsoc-2021-intro.md

+++
title = "Google Summer of Code 2021 Intro"
date = 2021-07-16
extra.author = "Adam Blanchet and Devin Ragotzy"
+++

Hello all! This year, Ruma has the pleasure of having two students for [Google

> Summer of Code (GSoC) 2021][gsoc]: Devin and Adam. We'll let them [re]introduce

Reviewer: Maybe make a small note so the `"newer"` below makes sense for everybody:

```suggestion
Summer of Code (GSoC) 2021][gsoc]: Devin (who worked on Ruma's proc-macro code during last year's GSoC) and Adam. We'll let them [re]introduce
```

## Identity Service API

Although I had made progress before the official coding period for GSoC had
started, I continued with finishing the remaining unimplemented Identity Service
API endpoints:

- [3PID validity endpoint][ruma/ruma#618]
- [3PID bind endpoint][ruma/ruma#621]
- [invitation signing endpoint][ruma/ruma#626]
- [store invitation endpoint][ruma/ruma#631]
- [3PID unbind endpoint][ruma/ruma#627]

Apart from the [3PID unbind endpoint][ruma/ruma#627] which is currently blocked,
this concludes the full coverage of the Identity Service API. This means that
now projects built upon Ruma will be able to more easily integrate functionality
with identity servers. This could also enable easier development of an identity
server written in Rust!

[ruma/ruma#618]: https://github.com/ruma/ruma/pull/618
[ruma/ruma#621]: https://github.com/ruma/ruma/pull/621
[ruma/ruma#626]: https://github.com/ruma/ruma/pull/626
[ruma/ruma#631]: https://github.com/ruma/ruma/pull/631
[ruma/ruma#627]: https://github.com/ruma/ruma/pull/627

## Implementing MSCs and adding types

I then moved on to a few other miscellaneous issues, to get accustomed to
working on the other parts of the Ruma project:

> - [add client secret and session ID identifier types][ruma/ruma#640]:

Reviewer: `ID identifier` is redundant (@Frinksy)

Should have noticed this in the PR title already ^^

Reviewer 2: ```suggestion
- [Add client secret and session identifier types][ruma/ruma#640]:
```

  I found this PR to be more interesting than its title may indicate. Not only
  did I add two identifier types, but I added a new macro to create validated

>   identifier types.The new macro was also used to greatly simplify the

Reviewer: ```suggestion
  identifier types. The new macro was also used to greatly simplify the
```

  declaration of `ServerName` struct. I really enjoyed working on this PR. It
  was an occasion to create a new macro to simplify the creation of these and

>   new types.  I had never written macro code before, and it was a good occasion

Reviewer: ```suggestion
  new types. I had never written macro code before, and it was a good occasion
```

  for me to learn something new!

> - [add RoomName struct][ruma/ruma#645]:

Reviewer: ```suggestion
- [Add RoomName struct][ruma/ruma#645]:
```

The PR title starts uppercase, I don't see why it shouldn't here. Same for the others.

  This one did not involve macro code, though similar in principle to the
  previous PR. It adds a `RoomName` struct to be used in an event, which
  enforces a maximum length.

> - [update endpoints for Blurhash implementation][ruma/ruma#650]:

Reviewer 2: ```suggestion
- [Update endpoints for Blurhash implementation][ruma/ruma#650]:
```

  [MSC2488][] adds the use of Blurhashes to the Matrix Spec. Although one
  `client-api` endpoint in Ruma had already been modified to add Blurhash

>   support, I extended this to all endpoints and events specified in the MSC.

Reviewer: I think the previous support didn't just affect a single endpoint. Maybe use some more general wording noting previous work on this in Ruma (@Frinksy)

Reviewer 2: ```suggestion
  [MSC2488][] adds the use of Blurhashes to the Matrix Spec. Although
  some Blurhash support had been previously added to Ruma, I
  extended this to all endpoints and events specified in the MSC.
```

> - [add "knock" feature from MSC2403][ruma/ruma#657]:

Reviewer 2: ```suggestion
- [Add "knock" feature from MSC2403][ruma/ruma#657]:
```

  This adds support for knocking on rooms, a feature which I'm looking forward
  to see implemented in clients. It was introduced in [MSC2403][].

> - [Implement reasons for leaving a room][ruma/ruma#663]:

Reviewer 2: ```suggestion
- [implement reasons for leaving a room][ruma/ruma#663]:
```
That's my bad since I missed it out originally, but I think it should remain consistent with the rest of the list.

> - [implement reasons for leaving a room][ruma/ruma#663]:

Reviewer 2: ```suggestion
- [Implement reasons for leaving a room][ruma/ruma#663]:
```

  Adding this feature, introduced in [MSC1983][], seemed rather straightforward.
  However, I bumped into an issue with a conditionally compiled field with a
  lifetime annotation. Thankfully, [@DevinR528][] helped diagnose the problem
  which was some macro code, for which they think they have found [a
  solution][ruma/ruma#664] along with [@jplatte][].

>   Once that fix is merged, we will also be able to merge my PR.

Reviewer: Small nit, this would make more sense to me (since there's more work to be done in my PR than just merge it).

```suggestion
  Once that fix is implemented, we will also be able to merge my PR.
```

[ruma/ruma#640]: https://github.com/ruma/ruma/pull/640
[ruma/ruma#645]: https://github.com/ruma/ruma/pull/645
[ruma/ruma#650]: https://github.com/ruma/ruma/pull/650
[MSC2488]: https://github.com/matrix-org/matrix-doc/blob/anoa/blurhash/proposals/2448-blurhash-for-media.md#profile-endpoints
[ruma/ruma#657]: https://github.com/ruma/ruma/pull/657
[MSC2403]: https://github.com/matrix-org/matrix-doc/blob/master/proposals/2403-knock.md
[ruma/ruma#663]: https://github.com/ruma/ruma/pull/663
[MSC1983]: https://github.com/matrix-org/matrix-doc/blob/master/proposals/1983-leave-reasons.md
[@DevinR528]: https://github.com/DevinR528
[ruma/ruma#664]: https://github.com/ruma/ruma/pull/664
[@jplatte]: https://github.com/jplatte

## What's next?

I'm looking forward to implementing the MSC for [Secure Secret Storage and

> Sharing][SSSS].  This is the second large part of my project, it will involve

Reviewer: ```suggestion
Sharing][SSSS]. This is the second large part of my project, it will involve
```

adding a number of events to support storing and sharing both keys and secrets.
To achieve full support, I would also need to implement crypto routines. I
expect that adding these will require more effort than adding the events, though
I look forward to the challenge.

I found my first half of GSoC to be very enriching. I really enjoyed working on
this project: I got to learn how to read and apply a specification, and I

> learned the rigour required for a project such as Ruma. I have sometimes been

Reviewer: ```suggestion
learned the rigor required for a project such as Ruma. I have sometimes been
```

We generally use American English.

caught off-guard: sometimes I miss extraneous newlines, sometimes (often) I
forget to add a changelog entry, and other similar more stylistic rather than
functional issues. Ruma is a big project with many crates, yet it maintains
consistency throughout, a quality I greatly appreciate. I now know that this
consistency comes from the thoroughness with which each change is checked, and I

> hope to improve my own rigour when it comes to each of my pull requests.

Reviewer: ```suggestion
hope to improve my own rigor when it comes to each of my pull requests.
```

- [check import rename][rust-lang/rust-clippy#7300]
- [check for disallowed types][rust-lang/rust-clippy#7315]

I [temporarily broke Clippy][rust-lang/rust-clippy#7434], Oh No! Luckily, the
fix seems to have worked.

- [fix the macro braces check][rust-lang/rust-clippy#7431]

I've spent a lot of time familiarizing myself with the rustc [usefulness][]
checker. The is how the Rust compiler determines if you have a match for all
possible patterns and checks if any of the patterns are useless.

```rust
let x = true;
match x { // `x` is matched exhaustively (all match patterns are useful)
    true => {},
    false => {},
    // true => {}, would be un-useful
}
```

This is where I now plan to add a lint to check for missed patterns on structs
and enums that are marked `non_exhaustive` ([see issue][rust-lang/rust#84332]).
This is helpful to Ruma because all of our public types are marked in this way,
allowing us to update our types in a non-breaking way while still informing
users of the change. Also, for anyone interested, I saw an interesting [blog
post][usefulness checking blog] about usefulness checking.

- [rustc reachable lint PR][rust-lang/rust#86809]

> For the last few days, I have been working with my mentor Jonas Platte to find

Reviewer: I was mentioned above with my username, now with full name. Might be good to have that consistent? I don't care that much either way though.
